Build with cflag `GC_FINAL_STATS` failed

Build with master (02f80c6)

I'm trying to build julia with option `GC_FINAL_STATS`.
https://github.com/JuliaLang/julia/blob/eeb0b6952a4caf5e9f1b4ae751adbabc2aae8e13/src/options.h#L67-L70

*Make.user*
```make
# valgrind
CFLAGS = -DMEMDEBUG -DMEMDEBUG2 -DGC_FINAL_STATS
JULIA_CPU_TARGET=generic
```


errors

```sh
~/jl/julia$ make
    CC src/gc-debug.o
/home/cyhan/jl/julia/src/gc-debug.c: In function 'gc_stats_pagetable0':
/home/cyhan/jl/julia/src/gc-debug.c:748:35: error: 'pagetable0_t' has no member named 'allocmap'
  748 |         uint32_t line = pagetable0->allocmap[pg_i] | pagetable0->freemap[pg_i];
      |                                   ^~
/home/cyhan/jl/julia/src/gc-debug.c:748:64: error: 'pagetable0_t' has no member named 'freemap'
  748 |         uint32_t line = pagetable0->allocmap[pg_i] | pagetable0->freemap[pg_i];
      |                                                                ^~
/home/cyhan/jl/julia/src/gc-debug.c: In function 'gc_stats_pagetable1':
/home/cyhan/jl/julia/src/gc-debug.c:762:35: error: 'pagetable1_t' has no member named 'allocmap0'
  762 |         uint32_t line = pagetable1->allocmap0[pg_i] | pagetable1->freemap0[pg_i];
      |                                   ^~
/home/cyhan/jl/julia/src/gc-debug.c:762:65: error: 'pagetable1_t' has no member named 'freemap0'
  762 |         uint32_t line = pagetable1->allocmap0[pg_i] | pagetable1->freemap0[pg_i];
      |                                                                 ^~
/home/cyhan/jl/julia/src/gc-debug.c: In function 'gc_stats_pagetable':
/home/cyhan/jl/julia/src/gc-debug.c:777:25: error: 'memory_map' undeclared (first use in this function)
  777 |         uint32_t line = memory_map.allocmap1[pg_i] | memory_map.freemap1[pg_i];
      |                         ^~~~~~~~~~
```

https://github.com/JuliaLang/julia/blob/eeb0b6952a4caf5e9f1b4ae751adbabc2aae8e13/src/gc.h#L248-L250

Related pr: #50137  @d-netto

--- a/src/gc-debug.c
+++ b/src/gc-debug.c
@@ -745,13 +745,11 @@ void gc_final_pause_end(int64_t t0, int64_t tend)
 static void gc_stats_pagetable0(pagetable0_t *pagetable0, unsigned *p0)
 {
     for (int pg_i = 0; pg_i < REGION0_PG_COUNT / 32; pg_i++) {
-        uint32_t line = pagetable0->allocmap[pg_i] | pagetable0->freemap[pg_i];
-        if (line) {
-            for (int j = 0; j < 32; j++) {
-                if ((line >> j) & 1) {
-                    (*p0)++;
-                }
-            }
+        uint8_t meta = pagetable0->meta[pg_i];
+        assert(meta == GC_PAGE_UNMAPPED || meta == GC_PAGE_ALLOCATED ||
+               meta == GC_PAGE_LAZILY_FREED || meta == GC_PAGE_FREED);
+        if (meta != GC_PAGE_UNMAPPED) {
+            (*p0)++;
         }
     }
 }
@@ -759,30 +757,24 @@ static void gc_stats_pagetable0(pagetable0_t *pagetable0, unsigned *p0)
 static void gc_stats_pagetable1(pagetable1_t *pagetable1, unsigned *p1, unsigned *p0)
 {
     for (int pg_i = 0; pg_i < REGION1_PG_COUNT / 32; pg_i++) {
-        uint32_t line = pagetable1->allocmap0[pg_i] | pagetable1->freemap0[pg_i];
-        if (line) {
-            for (int j = 0; j < 32; j++) {
-                if ((line >> j) & 1) {
-                    (*p1)++;
-                    gc_stats_pagetable0(pagetable1->meta0[pg_i * 32 + j], p0);
-                }
-            }
+        pagetable0_t *pagetable0 = pagetable1->meta0[pg_i];
+        if (pagetable0 == NULL) {
+            continue;
         }
+        (*p1)++;
+        gc_stats_pagetable0(pagetable0, p0);
     }
 }
 
 static void gc_stats_pagetable(unsigned *p2, unsigned *p1, unsigned *p0)
 {
     for (int pg_i = 0; pg_i < (REGION2_PG_COUNT + 31) / 32; pg_i++) {
-        uint32_t line = memory_map.allocmap1[pg_i] | memory_map.freemap1[pg_i];
-        if (line) {
-            for (int j = 0; j < 32; j++) {
-                if ((line >> j) & 1) {
-                    (*p2)++;
-                    gc_stats_pagetable1(memory_map.meta1[pg_i * 32 + j], p1, p0);
-                }
-            }
+        pagetable1_t *pagetable1 = alloc_map.meta1[pg_i];
+        if (pagetable1 == NULL) {
+            continue;
         }
+        (*p2)++;
+        gc_stats_pagetable1(pagetable1, p1, p0);
     }
 }
 
@@ -791,7 +783,7 @@ void jl_print_gc_stats(JL_STREAM *s)
 #ifdef _OS_LINUX_
     malloc_stats();
 #endif
-    double ptime = jl_clock_now() - process_t0;
+    double ptime = jl_hrtime() - process_t0;
     jl_safe_printf("exec time\t%.5f sec\n", ptime);
     if (gc_num.pause > 0) {
         jl_safe_printf("gc time  \t%.5f sec (%2.1f%%) in %d (%d full) collections\n",
@@ -1012,7 +1004,7 @@ void jl_gc_debug_init(void)
 #endif
 
 #ifdef GC_FINAL_STATS
-    process_t0 = jl_clock_now();
+    process_t0 = jl_hrtime();
 #endif
 }
 

--- a/src/gc-debug.c
+++ b/src/gc-debug.c
@@ -744,7 +744,7 @@ void gc_final_pause_end(int64_t t0, int64_t tend)
 
 static void gc_stats_pagetable0(pagetable0_t *pagetable0, unsigned *p0)
 {
-    for (int pg_i = 0; pg_i < REGION0_PG_COUNT / 32; pg_i++) {
+    for (int pg_i = 0; pg_i < REGION0_PG_COUNT; pg_i++) {
         uint8_t meta = pagetable0->meta[pg_i];
         assert(meta == GC_PAGE_UNMAPPED || meta == GC_PAGE_ALLOCATED ||
                meta == GC_PAGE_LAZILY_FREED || meta == GC_PAGE_FREED);
@@ -756,7 +756,7 @@ static void gc_stats_pagetable0(pagetable0_t *pagetable0, unsigned *p0)
 
 static void gc_stats_pagetable1(pagetable1_t *pagetable1, unsigned *p1, unsigned *p0)
 {
-    for (int pg_i = 0; pg_i < REGION1_PG_COUNT / 32; pg_i++) {
+    for (int pg_i = 0; pg_i < REGION1_PG_COUNT; pg_i++) {
         pagetable0_t *pagetable0 = pagetable1->meta0[pg_i];
         if (pagetable0 == NULL) {
             continue;
@@ -768,7 +768,7 @@ static void gc_stats_pagetable1(pagetable1_t *pagetable1, unsigned *p1, unsigned
 
 static void gc_stats_pagetable(unsigned *p2, unsigned *p1, unsigned *p0)
 {
-    for (int pg_i = 0; pg_i < (REGION2_PG_COUNT + 31) / 32; pg_i++) {
+    for (int pg_i = 0; pg_i < REGION2_PG_COUNT; pg_i++) {
         pagetable1_t *pagetable1 = alloc_map.meta1[pg_i];
         if (pagetable1 == NULL) {
             continue;

--- a/src/gc-pages.c
+++ b/src/gc-pages.c
@@ -100,7 +100,7 @@ NOINLINE jl_gc_pagemeta_t *jl_gc_alloc_page(void) JL_NOTSAFEPOINT
     // try to get page from `pool_lazily_freed`
     meta = pop_lf_page_metadata_back(&global_page_pool_lazily_freed);
     if (meta != NULL) {
-        gc_alloc_map_set(meta->data, 1);
+        gc_alloc_map_set(meta->data, GC_PAGE_ALLOCATED);
         // page is already mapped
         return meta;
     }
@@ -108,14 +108,14 @@ NOINLINE jl_gc_pagemeta_t *jl_gc_alloc_page(void) JL_NOTSAFEPOINT
     // try to get page from `pool_clean`
     meta = pop_lf_page_metadata_back(&global_page_pool_clean);
     if (meta != NULL) {
-        gc_alloc_map_set(meta->data, 1);
+        gc_alloc_map_set(meta->data, GC_PAGE_ALLOCATED);
         goto exit;
     }
 
     // try to get page from `pool_freed`
     meta = pop_lf_page_metadata_back(&global_page_pool_freed);
     if (meta != NULL) {
-        gc_alloc_map_set(meta->data, 1);
+        gc_alloc_map_set(meta->data, GC_PAGE_ALLOCATED);
         goto exit;
     }
 
@@ -155,7 +155,7 @@ exit:
 void jl_gc_free_page(jl_gc_pagemeta_t *pg) JL_NOTSAFEPOINT
 {
     void *p = pg->data;
-    gc_alloc_map_set((char*)p, 0);
+    gc_alloc_map_set((char*)p, GC_PAGE_FREED);
     // tell the OS we don't need these pages right now
     size_t decommit_size = GC_PAGE_SZ;
     if (GC_PAGE_SZ < jl_page_size) {

--- a/src/gc.c
+++ b/src/gc.c
@@ -1448,6 +1448,7 @@ done:
             push_lf_page_metadata_back(&global_page_pool_freed, pg);
         }
         else {
+            gc_alloc_map_set(pg->data, GC_PAGE_LAZILY_FREED);
             push_lf_page_metadata_back(&global_page_pool_lazily_freed, pg);
         }
     #else

--- a/src/gc.h
+++ b/src/gc.h
@@ -257,6 +257,11 @@ typedef struct {
     pagetable1_t *meta1[REGION2_PG_COUNT];
 } pagetable_t;
 
+#define GC_PAGE_UNMAPPED        0
+#define GC_PAGE_ALLOCATED       1
+#define GC_PAGE_LAZILY_FREED    2
+#define GC_PAGE_FREED           3
+
 extern pagetable_t alloc_map;
 
 STATIC_INLINE uint8_t gc_alloc_map_is_set(char *_data) JL_NOTSAFEPOINT
@@ -272,7 +277,7 @@ STATIC_INLINE uint8_t gc_alloc_map_is_set(char *_data) JL_NOTSAFEPOINT
     if (r0 == NULL)
         return 0;
     i = REGION0_INDEX(data);
-    return r0->meta[i];
+    return (r0->meta[i] == GC_PAGE_ALLOCATED);
 }
 
 STATIC_INLINE void gc_alloc_map_set(char *_data, uint8_t v) JL_NOTSAFEPOINT

